src/main/scala/edu/cmu/cs/obsidian/typecheck/Error.scala

```diff
@@ -360,3 +361,94 @@ case class ConstructorAnnotationMissingError(contractName: String) extends Error
+                                     interfaceRetType: Option[ObsidianType]) extends Error {
+    val msg: String =
+        if (retType.isDefined) {
+            s"Transaction $transName returns ${retType.get}, but the transaction it implements has no return type."
+        } else {
+            s"Transaction $transName has no return type, but the transaction it implements returns ${interfaceRetType.get}."
+        }
+}
+
+case class MissingStateImplError(contractName: String, interfaceName: String, stateName: String) extends Error {
+    val msg: String =
+        s"Missing state $stateName from interface $interfaceName in contract $contractName."
+}
+
+case class MissingTransactionImplError(contractName: String, interfaceName: String,
+                                       transactionName: String) extends Error {
+    val msg: String =
+        s"Missing transaction $transactionName from interface $interfaceName in contract $contractName."
+}
+
+case class GenericParameterListError(paramsLength: Int, actualLength: Int) extends Error {
+    val msg: String =
+        s"Expected $paramsLength generic parameters but got $actualLength parameters"
+}
+
+case class GenericParameterError(param: GenericType, actualArg: ObsidianType) extends Error {
+    val msg: String =
+        s"Argument $actualArg is not compatible with generic parameter $param"
+}
+
+case class GenericParameterAssetError(paramName: String, argTypeName: String) extends Error {
+    val msg: String =
+        s"Argument type $argTypeName is an asset, but the type parameter $paramName is not allowed to be an asset. Add the 'asset' modifier to $paramName to allow the parameter to be an asset."
+}
+
+case class NonPrimitiveTypeTransformError(originalType: NonPrimitiveType, transformedType: ObsidianType) extends Error {
+    override val msg: String =
+        s"Type $originalType was transformed into $transformedType, which isn't a non-primitive type."
+}
+
+case class PermissionCheckRedundant(actualPerm: Permission, testPermission: Permission,
+                                    isSubperm: Boolean) extends Error {
+    private val isSubpermStr: String =
+        if (isSubperm) {
+            "is a subpermission. The then-branch will always be taken"
+        } else {
+            "is not a subpermission. The else-branch will always be taken"
+        }
+
+    override val msg: String =
+        s"Redundant permission check. Testing for permission $testPermission, but actual permission is $isSubperm, " +
+            s"which $isSubpermStr."
+}
+
+case class AssetStateImplError(implState: State, interfaceState: State) extends Error {
+    override val msg: String =
+        s"State ${implState.name} is an asset state but it implements a state which is not an asset state."
+}
+
+case class BadFFIInterfaceBoundError(name: String) extends Error {
+    override val msg: String =
+        s"Cannot override '$name' because it is an FFI contract."
```

Comment:
Not sure if users will know what "FFI" means. Maybe "because it is implemented in Java, not Obsidian."